### DISCUSSION TITLE
MQTT publishing to freakent / dbus-mqtt-devices

### DISCUSSION THREAD

**Discussion by stromibaer** [tbnobody/OpenDTU#217] [Ideas]
Hallo,
wäre es möglich, wie beim Home Assistent ein eigenes MQTT modul zu ergänzen, welches die Topics und Werte entsprechend dem Inputformat vom DBUS-MQTT für PV-Inverter entspricht. Damit könnten die Werte direkt an Victron Venus/Cerbo MQTT Broker gesendet werden und der Hoymiles Inverter würde mit der Ergänzung von Freakent am DBUS registriert werden und im VRM Portal mit den Werten ersichtlich sein.

Unter: https://github.com/freakent/dbus-mqtt-devices/blob/main/test-data/pvinverter-simulator.py
Ist ein Beispiel, welche Werte mit Topic zu senden sind und wie der Ablauf ist.

Danke,
Peter

**Comment by tbnobody** [comment]
Hallo Peter,

wenn ich das richtig sehe, wird da ähnlich wie bei dem Home Assistant Auto Discovery einfach eine Sub-Topic im Broker mit eine Struktur gefüllt die beschreibt wo welche Werte stehen. Muss ich mir mal genauer anschauen. Klingt jetzt aber nicht soo komplex.

**Reply by stromibaer** [reply]
Eine erste Draftversion davon habe ich schon fertig geschrieben, wobei ich bin kein Programmierer, eher copy/paster. Soweit funktioniert der Senderteil, am Victron werden mir auch die Werte angezeigt. Einzig im VRM Portal und im Cerbos bekomme ich die Inverter noch nicht angezeit. Ist aber irgendwo ein Fehler auf der Victronseite.

![image](https://user-images.githubusercontent.com/47776842/196387005-b62d89ac-a09f-49df-bae8-bbfa54b9097d.png)

![image](https://user-images.githubusercontent.com/47776842/196387111-fccd5b03-abb8-4586-9bc7-07aab0967141.png)

![image](https://user-images.githubusercontent.com/47776842/196387318-46497061-de1e-450a-bbbb-562ad56f1d0f.png)

**Reply by dersch81** [reply]
Sehr geil!

**Reply by dersch81** [reply]
Kannst du deinen ersten Draft bereit stellen? Dann kann der ein oder andere auch mitarbeiten :)

**Reply by stromibaer** [reply]
Gerne, wie gesagt, betrete mit Git und programmieren etwas Neuland. Hab nun einen Fork erstellt: 
[https://github.com/stromibaer/OpenDTU-Victron](url)
und die Änderungen dort reingeladen. Im wesentlichen in den beiden Dateien mqttpubishing und mqttsettings, sowie der libs dazu.

Offen sind noch folgende Themen:
1) Runden der Ausgabewerte --> Mit `fieldvalue = round(fieldvalue * 100)/100;` bekomme ich das zwar hin, aber es wird 1,9100000000081 ausgegeben.

2) Senden eines "Disconnects" wenn MqTT beendet wird bzw. auch wenn der Inverter keine Werte sendet --> Victron Venus nimmt die letzten MqTT Werte her und rechnet mit diesen weiter.

3) Um die anderen Interessenten nicht mit dem Teil zu "belästigen" einen Menüknopf für "enablen/disablen" des MqTT publishen der Werte für Victron Venus OS, dazu muss allerdings in der WebApp angepasst werden und in der Config ein zusätzlicher Parameter eingefügt werden --> Ist etwas mehr Aufwand und ich noch nicht dahintergestiegen, wie ich die WebApp korrekt kompiliere, hab diese mit Yarn bis jetzt versucht.

4) Dauertest ist noch ausständig, was das Victron Venus OS über Nacht macht, wenn der Inverter nicht sendet, sollte mit Punkt 2 abgefangen werden, sonst kommt nur Unfug raus bei den Statistiken.

Sonst fällt mir nix ein, getestet bis jetzt mit Standard ESP32 und Olimex.

Die services.yml Datei von [https://github.com/freakent/dbus-mqtt-devices](url) hab ich noch angepasst, damit man den Inverter entweder auf AC-IN oder AC-OUT umstellen kann, je nachdem, wo dieser einspeist. Dazu stelle ich noch einen Request, damit die Änderungen in den Release aufgenommen werden.

**Reply by dersch81** [reply]
Was du da machst ist genau das was ich suche! Ich glaube auch. das wird generell OpenDTU auf ein weiteres Level bringen da eben auch alle Victron ESS Bastler (so wie ich, im Aufbau) auch gerne Hoymiles einsetzen. Da treffen sich zwei kompatible Welten. Ich bin auch kein Programmierer und werde wohl wenig beitragen können :( Es wäre so toll wenn das Upstream wird :)

**Reply by stromibaer** [reply]
![image](https://user-images.githubusercontent.com/47776842/197131883-0d58f1ef-77f3-453a-b11e-80b87a5db4b6.png)

**Reply by dersch81** [reply]
Ich habe mir jetzt mal deinen Fork gebaut und geflasht. Aber ich kapiere nicht wie und wo ich ihm den Weg zum VenusOS zeige.

**Reply by stromibaer** [reply]
Unter Settings -> General -> SSH am Cerbo aktivieren und Root Passwort vergeben:
![image](https://user-images.githubusercontent.com/47776842/197271550-4d37e5ec-ca3f-44df-9bca-3ce1ed54ea81.png)

Unter Settings -> Services -> MqTT aktivieren:
![image](https://user-images.githubusercontent.com/47776842/197272239-be0bfc49-4219-4143-847e-eeeceabf02c3.png)

Dann über SSH einloggen am Cerbo und die Software von [https://github.com/freakent/dbus-mqtt-devices](url) installieren.
Falls noch nicht im Mainstream eingepflegt, die servics.yml Datei anpassen, gemäß [https://github.com/freakent/dbus-mqtt-devices/issues/23](url) --> Die Datei liegt unter /data/drivers/dbus-mqtt-devices
![image](https://user-images.githubusercontent.com/47776842/197272505-90742086-9897-4d00-a62a-59c8269b25e8.png)

Installationsanleitung für dbus-mqtt-devices ist eh dabei. Nach Reboot vom Cerbo in der Open-DTU die IP-Adresse vom Cerbo als MqTT Zielserver eintragen: 
![image](https://user-images.githubusercontent.com/47776842/197272705-c685caf0-3f28-4bb4-89df-374722b56fff.png)

Mit dem MqTT Explorer kann man prüfen, was so abgeht am MqTT Server, dazu als Zieladresse wieder die IP-Adresse vom Cerbo eingeben. Es sollte ein Status Eintrag unter Devices mit HM+Seriennummer des Inverters von der Open-DTU geschrieben werden:
![image](https://user-images.githubusercontent.com/47776842/197273021-6508ddd1-3d8f-438c-8bbd-6236e0172e9b.png)

Als Antwort sendet Venus OS die Portal-Id und erstellt für den Inverter einen Deviceeintrag im VRM-Portal. Dieses liest die Open-DTU wieder aus und sendet unter W/Portal-Id/.... die entsprechenden Werte:
![image](https://user-images.githubusercontent.com/47776842/197273408-29f94d2d-735b-4391-a867-e525ed057aea.png)

Falls etwas schiefgeht, dann erscheint unter W/NOportalid ... NOdeviceInstance, dann den MqTT Server an der Open-DTU disablen und wieder enablen. Die Regiestrierung erfolgt nur einmalig --> Bis jetzt zumindest, könnte man noch in einer Fehlerbehandlung erweitern.

Wenn der Inverter im Menu am Cerbo auftaucht, kann man im Setup noch AC-IN vss. AC-OUT ändern, je nachdem wo der Inverter einspeist.

Aktuell wird alles Phase L1 dargestellt, es gibt aber schon eine Anfrage es dreiphasig darzustellen. Manche haben mehrere Inverter auf unterschiedlichen Phasen L1,L2 und L3 aufgeteilt. Da müsste man aber am Besten die jeweilige Phase im Invertermenü von der Open-DTU eintragen können, dann kann man diese dort auslesen und die Werte entsprechend über MqTT mit der richtigen Phase publishen.

**Comment by dersch81** [comment]
Da ich mir grade ein Victron ESS aufbaue mit VenusOS wäre ich da auch sehr dankbar eine Möglichkeit zu haben. Ich dachte auch schon via NodeRed die Daten an dbus-mqtt zu senden. Verstehe dbus-mqtt aber noch nicht so ganz.
